File: kommandant-core/src/main/java/com/github/kvnxiao/kommandant/command/CommandContext.java
package com.github.kvnxiao.kommandant.command;

import org.jetbrains.annotations.Nullable;

/**
 * A data class for containing the context of the command. It parses a string and splits it in two based on the first
 * available space. The first string represents the alias, and the second string represents the arguments leftover.
 * <p>
 * NOTE: This class was written in Java. And why, you might ask? As of Kotlin 1.10, there is still some performance
 * overhead in splitting strings by a literal. As such, string splitting in Java is currently still more performant
 * than in Kotlin.
 */
public class CommandContext {

    /**
     * The first string after splitting the input string.
     */
    private final String alias;

    /**
     * The second string after splitting the input string.
     */
    private final String args;

    /**
     * The command associated with this context.
     */
    private ICommand command = null;

    /**
     * The string literal to split the incoming input string by, defined as a single space character " ".
     */
    private static final String SPLIT_REGEX = " ";

    /**
     * Constructor which takes in a string to split in two. Splits based on the first available space character.
     * If a space does not exist, the args string will be null. If the input string is null, both resultant args and
     * aliases will be null.
     *
     * @param message The message string to split into alias and args.
     */
    public CommandContext(@Nullable final String message) {
        if (message != null) {
            final String[] context = message.split(SPLIT_REGEX, 2);
            this.alias = (context.length > 0) ? context[0] : null;
            this.args = (context.length == 2) ? context[1] : null;
        } else {
            this.args = null;
            this.alias = null;
        }
    }

    /**
     * Secondary constructor which sets the alias and args of the context based on the provided input parameters.
     *
     * @param alias The alias for the command context.
     * @param args The arguments for the command context.
     */
    public CommandContext(final String alias, final String args) {
        this.alias = alias;
        this.args = args;
    }

    /**
     * Gets the alias for this CommandContext. Can be null.
     *
     * @return The (nullable) alias.
     */
    @Nullable
    public String getAlias() {
        return alias;
    }

    /**
     * Gets the args for this CommandContext. Can be null.
     *
     * @return The (nullable) args.
     */
    @Nullable
    public String getArgs() {
        return args;
    }

    /**
     * A CommandContext has an alias if the alias string is not null and not empty.
     *
     * @return Whether an alias exists.
     */
    public boolean hasAlias() {
        return alias != null && !alias.isEmpty();
    }

    /**
     * A CommandContext has args if the args string is not null. The args can be null if the input string that was
     * split had no space character in it.
     *
     * @return Whether args exist.
     */
    public boolean hasArgs() {
        return args != null;
    }

    /**
     * Gets the command this context is associated with.
     *
     * @return The associated command for this context.
     */
    @Nullable
    public ICommand getCommand() {
        return this.command;
    }

    /**
     * Sets the command this context is associated with.
     *
     * @param command The command to associate with this context.
     */
    public void setCommand(ICommand command) {
        this.command = command;
    }

}

File: kommandant-core/src/main/kotlin/com/github/kvnxiao/kommandant/command/CommandContext.kt
/*
 * Copyright 2017 Ze Hao Xiao
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
package com.github.kvnxiao.kommandant.command

/**
 * A data class for holding the context of the command. It parses a string and splits it in two based on the first
 * available space. The first string represents the alias, and the second string represents the arguments leftover.
 *
 * @property[alias] The alias or prefixed-alias used to call the command.
 * @property[args] The (nullable) arguments provided for the command.
 * @property[properties] The properties of the command being called.
 */
open class CommandContext(
    val alias: String,
    val args: String?,
    val properties: CommandProperties) {

    constructor(alias: String, args: String?, command: ICommand<*>): this(alias, args, command.props)

    /**
     * Whether the command context contains arguments for the command.
     *
     * @return[Boolean] Whether the args property in the command context is not null.
     */
    fun hasArgs(): Boolean = args !== null

    override fun equals(other: Any?): Boolean {
        if (this === other) return true
        if (other !is CommandContext) return false

        if (alias != other.alias) return false
        if (args != other.args) return false
        if (properties != other.properties) return false

        return true
    }

    override fun hashCode(): Int {
        var result = alias.hashCode()
        result = 31 * result + (args?.hashCode() ?: 0)
        result = 31 * result + properties.hashCode()
        return result
    }

    override fun toString(): String {
        return "CommandContext(alias='$alias', args=$args, properties=$properties)"
    }

}